Framework/CurveFitting/src/Functions/InelasticDiffRotDiscreteCircle.cpp

```diff
@@ -145,21 +142,21 @@ void InelasticDiffRotDiscreteCircle::setWorkspace(
     return;
 
   size_t numHist = workspace->getNumberHistograms();
+  const auto &spectrumInfo = workspace->spectrumInfo();
   for (size_t idx = 0; idx < numHist; idx++) {
-    Mantid::Geometry::IDetector_const_sptr det;
-    try {
-      det = workspace->getDetector(idx);
-    } catch (Kernel::Exception::NotFoundError &) {
+    if (!spectrumInfo.hasDetectors(idx)) {
       m_qValueCache.clear();
       g_log.information("Cannot populate Q values from workspace");
       break;
     }
 
+    const auto &det = spectrumInfo.detector(idx);
+
     try {
-      double efixed = workspace->getEFixed(det);
-      double usignTheta = 0.5 * workspace->detectorTwoTheta(*det);
+      double efixed = workspace->getEFixed(det.getID());
+      double usingTheta = 0.5 * workspace->detectorTwoTheta(det);
```

Comment:
Would be good not to go through `IDetector` at all of possible. Can this be done through `SpectrumInfo` instead?